tools/manifest/vcs.py

```diff
@@ -70,27 +75,168 @@ def __iter__(self):
+        mtime_cache = self.mtime_cache
+        for dirpath, dirnames, filenames in self.path_filter(walk(".")):
+            for filename, path_stat in filenames:
+                # We strip the ./ prefix off the path
+                path = os.path.join(dirpath, filename)
+                if mtime_cache is None or mtime_cache.updated(path, path_stat):
+                    yield SourceFile(self.root, path, self.url_base), True
+                else:
+                    yield path, False
+        self.ignore_cache.dump()
+
+    def dump_caches(self):
+        for cache in [self.mtime_cache, self.ignore_cache]:
+            if cache is not None:
+                cache.dump()
+
+
+class CacheFile(object):
+    file_name = None
+
+    def __init__(self, cache_root, rebuild=False):
+        if not os.path.exists(cache_root):
+            os.makedirs(cache_root)
+        self.path = os.path.join(cache_root, self.file_name)
+        self.data = self.load(rebuild)
+        self.modified = False
+
+    def dump(self):
+        if not self.modified:
+            return
+        with open(self.path, 'w') as f:
+            json.dump(self.data, f, indent=1)
+
+    def load(self, rebuild=False):
+        data = {}
+        try:
+            if not rebuild:
+                with open(self.path, 'r') as f:
+                    data = json.load(f)
+                data = self.check_valid(data)
+        except IOError:
+            pass
+        return data
+
+    def check_valid(self, data):
+        """Check if the cached data is valid and return an updated copy of the
+        cache containing only data that can be used."""
+        return data
+
+
+class MtimeCache(CacheFile):
+    file_name = "mtime.json"
+
+    def updated(self, rel_path, stat):
+        """Return a boolean indicating whether the file changed since the cache was last updated.
+
+        This implicitly updates the cache with the new mtime data."""
+        mtime = stat.st_mtime
+        if mtime != self.data.get(rel_path):
+            self.modified = True
+            self.data[rel_path] = mtime
+            return True
+        return False
+
+
+class GitIgnoreCache(CacheFile):
+    file_name = "gitignore.json"
+
+    def __init__(self, cache_root, ignore_path, rebuild=False):
+        self.ignore_path = ignore_path
+        super(GitIgnoreCache, self).__init__(cache_root, rebuild=False)
+
+    def check_valid(self, data):
+        mtime = os.path.getmtime(self.ignore_path)
+        if data.get("/gitignore_file") != [self.ignore_path, mtime]:
+            self.modified = True
+            data = {}
+            data["/gitignore_file"] = [self.ignore_path, mtime]
+        return data
+
+    def __contains__(self, key):
+        return key in self.data
+
+    def __getitem__(self, key):
+        return self.data[key]
+
+    def __setitem__(self, key, value):
+        if self.data.get(key) != value:
+            self.modified = True
+            self.data[key] = value
+
+
+def walk(root):
```

Comment:
Just because it's one less required dependency, and a C one at that. Also since we typically require the stat data anyway it's not clear how much of a win it will be.